src/main/scala/mesosphere/marathon/api/akkahttp/v2/TasksController.scala

```diff
@@ -0,0 +1,105 @@
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.api.akkahttp.Controller
+import mesosphere.marathon.api.akkahttp.Directives.pathEndOrSingleSlash
+import mesosphere.marathon.core.appinfo.EnrichedTask
+import mesosphere.marathon.core.condition.Condition
+import mesosphere.marathon.core.group.GroupManager
+import mesosphere.marathon.core.health.{ Health, HealthCheckManager }
+import mesosphere.marathon.core.instance.Instance
+import mesosphere.marathon.core.instance.Instance.Id
+import mesosphere.marathon.core.task.tracker.InstanceTracker
+import mesosphere.marathon.plugin.auth.{ Authenticator, _ }
+import play.api.libs.json.Json
+import mesosphere.marathon.raml.{ AnyToRaml, EnrichedTasksList, Writes }
+import mesosphere.marathon.state.PathId
+
+import scala.async.Async._
+import scala.concurrent.{ ExecutionContext, Future }
+
+class TasksController(instanceTracker: InstanceTracker, groupManager: GroupManager, healthCheckManager: HealthCheckManager)(
+    implicit
+    val actorSystem: ActorSystem,
+    val executionContext: ExecutionContext,
+    val authenticator: Authenticator,
+    val authorizer: Authorizer
+) extends Controller with StrictLogging {
+
+  import mesosphere.marathon.api.akkahttp.Directives._
+  import mesosphere.marathon.api.akkahttp.EntityMarshallers._
+  import mesosphere.marathon.raml.EnrichedTaskConversion._
+
+  override val route = {
+    get {
+      pathEndOrSingleSlash {
+        listTasks()
+      }
+    }
+  }
+
+  private def listTasks(): Route = {
+    authenticated.apply { implicit identity =>
+      authorized(ViewResource, AuthorizedResource.SystemConfig).apply {
+        parameters("status".?, "status[]".as[String].*) { (statusParameter, statusesParameter) =>
+          val statuses = statusParameter.map(s => Seq(s) ++ statusesParameter).getOrElse(Seq[String](statusesParameter.toSeq: _*))
+          onSuccess(enrichedTasks(statuses)) { tasks =>
+            complete(TasksList(tasks).toRaml)
+          }
+        }
+      }
+    }
+  }
+
+  @SuppressWarnings(Array("all")) // async/await
+  private def enrichedTasks(statuses: Seq[String]): Future[Seq[EnrichedTask]] = async {
+    val conditionSet: Set[Condition] = statuses.flatMap(toTaskState)(collection.breakOut)
+
+    val instancesBySpec = await(instanceTracker.instancesBySpec)
+
+    val instances: Iterable[(PathId, Instance)] = instancesBySpec.instancesMap.values.flatMap { appTasks =>
+      appTasks.instances.map(i => appTasks.specId -> i)
+    }
+    val appIds: Set[PathId] = instancesBySpec.allSpecIdsWithInstances
+
+    val appToPorts: Map[PathId, Seq[Int]] = groupManager.apps(appIds).map {
+      case (appId, app) => appId -> app.map(_.servicePorts).getOrElse(Nil)
+    }
+
+    val instancesHealth = await(
+      Future.sequence(appIds.map { appId =>
+        healthCheckManager.statuses(appId)
+      })).foldLeft(Map[Id, Seq[Health]]())(_ ++ _)
+
+    val enrichedTasks: Iterable[Iterable[EnrichedTask]] = for {
```

Comment:
We try to avoid for comprehensions for performance reasons.

Reply:
hmm, never heard of that (not saying it is not true). Can you maybe explain more or provide some article... ?

Reply:
I could not find the source anymore. I guess overall the downside is that it's hiding performance bottlenecks.

Reply:
Sorry, I could not find the source. I guess in general it's hiding bottlenecks. E.g. this could be:

```scala
instances
  .filter( (appId, instance) => conditionSet.isEmpty || conditionSet(instance.state.condition))
  .flatMap( (appId, instance)  =>
    tasks = instance.tasksMap.values
    tasks.map { task => EnrichedTask(...)
  }(breakOut)
```

As you can see I already have issues placing the if-clause. The for comprehension is not really making this clear. A collect should work as well.